### DISCUSSION TITLE
specie block in input file

### DISCUSSION THREAD

**Discussion by Sufihoo** [epochpic/epoch#738] [Q&A]
I want to add a distribution function "dist_fn" which has two dirac delta functions as shown in attached figure
![delta function](https://github.com/user-attachments/assets/59f4d849-5e4c-4ae0-b0a7-5d5a6b8c4fb4)
. My question is how can we write a delta function in input.deck file as i see no such function in user manual?

**Comment by Status-Mirror** [comment]
Hi @Sufihoo,

You can set momenta distributions for particles using the method outlined in our [user distribution demo](https://epochpic.github.io/quickstart/basic_examples/particle_load.html#momentum-distributions).

Unfortunately, due to the way EPOCH samples these momenta distributions, a Dirac delta function would not be viable. EPOCH randomly samples `px`, `py` and `pz` values, and then tests whether these satisfy the momentum distribution - a Dirac delta function would make this incredibly rare, as having both $p_y$ and $p_z$ satisfy $\delta(v_\perp - v_{\perp 0})$ by chance would be near impossible.

You could try sampling momenta externally, along with particle positions, and load the particles like in our [particles from file demo](https://epochpic.github.io/quickstart/basic_examples/particle_load.html#load-particles-from-file), but this could be a long process to generate these files.

If you're comfortable coding in FORTRAN, you could modify the source-code to change the sampling behaviour. We don't generally provide help for editing the source-code, but here's an idea of how I would go about it:

Let's say we wanted to modify `epoch2d`. There is a subroutine called `sample_from_deck_expression` in `epoch/epoch2d/src/user_interaction/particle_temperature.F90`. In this subroutine, there are lines of the form:

```
      parameters%pack_p(1) = random() * range_diff1 + ranges(1,1)
      parameters%pack_p(2) = random() * range_diff2 + ranges(2,1)
      parameters%pack_p(3) = random() * range_diff3 + ranges(3,1)
```

These lines are responsible for randomly sampling momenta from your ranges. These are then substituted into the distribution function to determine whether you they are kept for the current particle. Here, they're randomly sampled from within the $x$, $y$ and $z$ ranges you provide. If you need the perpendicular component to always have the same magnitude, then you could sample a magnitude and angle instead

```
      parameters%pack_p(1) = random() * range_diff1 + ranges(1,1)
      p_mag = random() * range_diff2 + ranges(2,1)
      p_theta = 2.0_num * pi * random()
      parameters%pack_p(2) = p_mag * COS(p_theta) 
      parameters%pack_p(3) = p_mag * SIN(p_theta)
```

where you declare p_mag and p_theta as REAL(num) variables. This would use the range given to `py` as your magnitude.

Hope this helps give you some ideas,
Stuart

**Reply by Sufihoo** [reply]
@Status-Mirror , thank you so much for your detailed answer. I just want to ask if I can set up a delta function by extending the maths parser as explained here [https://epochpic.github.io/developer/input_output/custom_maths_parser.html]?
The other thing is if i modify the source code as you said, how do i write distribution function (dist_fn = (dist_fn_max * delta(pz momentum_par) * delta(sqrt(px^2 + py^2)-momentum_perp))/ dist_fn_max) in input.deck file?

**Reply by Status-Mirror** [reply]
I don't think adding a delta-function to the maths parser will help you. The code will still attempt to sample `px`, `py`, `pz` values between the limits you give it. There is almost zero chance the code will randomly pick a `py` and `pz` which satisfies a dirac delta function, so the code will never stop trying to sample your distribution.

The only alternatives are to change the way the code samples `px`, `py` and `pz`, to hard-code a dirac-delta, or to externally calculate `px`, `py` and `pz` values for each particle, and load them from file as in the [particles from file demo](https://epochpic.github.io/quickstart/basic_examples/particle_load.html#load-particles-from-file).

We don't generally assist users trying to modify the source-code, so I would probably advise you to try the particles-from-file workaround. This will be easier if you're less familiar with FORTRAN coding, as modifying the momentum sampling behaviour internally would require a fair amount of coding.
